pilot/pkg/proxy/envoy/v2/discovery.go

```diff
@@ -377,48 +374,11 @@ func (s *DiscoveryServer) ConfigUpdate(full bool) {
 // It ensures that at minimum minQuiet time has elapsed since the last event before processing it.
 // It also ensures that at most maxDelay is elapsed between receiving an event and processing it.
 func (s *DiscoveryServer) handleUpdates(stopCh <-chan struct{}) {
-	var timeChan <-chan time.Time
-	var startDebounce time.Time
-	var lastConfigUpdateTime time.Time
-
-	pushCounter := 0
-
-	debouncedEvents := 0
-	fullPush := false
 
 	for {
 		select {
-		case r := <-s.updateChannel:
-			lastConfigUpdateTime = time.Now()
-			if debouncedEvents == 0 {
-				timeChan = time.After(DebounceAfter)
-				startDebounce = lastConfigUpdateTime
-			}
-			debouncedEvents++
-			// fullPush is sticky if any debounced event requires a fullPush
-			if r.full {
-				fullPush = true
-			}
-
-		case now := <-timeChan:
-			timeChan = nil
-
-			eventDelay := now.Sub(startDebounce)
-			quietTime := now.Sub(lastConfigUpdateTime)
-			// it has been too long or quiet enough
-			if eventDelay >= DebounceMax || quietTime >= DebounceAfter {
-				pushCounter++
-				adsLog.Infof("Push debounce stable[%d] %d: %v since last change, %v since last push, full=%v",
-					pushCounter, debouncedEvents,
-					quietTime, eventDelay, fullPush)
-
-				go s.doPush(fullPush)
-				fullPush = false
-				debouncedEvents = 0
-				continue
-			}
-
-			timeChan = time.After(DebounceAfter - quietTime)
+		case f := <-s.updateChannel.C:
+			f(s.doPush)
```

Comment:
Is doPush now blocking?

Reply:
Yes exactly. Running this in parallel caused the memory issue. The calculation of the PushContext  in #14782 took 5 to 10 seconds but the changes in k8s occurred every 2 seconds.
Therefore, the number of goroutines calculating PushContexts has grown over time without limit.

Reply:
Push context calculation took 10s? Do you have some profile info on the slowest function?

Reply:
We have done some profiling on a similar scenario. Here are the outputs:
![image](https://user-images.githubusercontent.com/13213351/62031492-cb6d2400-b1e7-11e9-8b38-40ae1c332226.png)